### ISSUE TITLE
Grapheme Breaks

### ISSUE THREAD

**Issue by roever** [adah1972/libunibreak#17]
Hi, have you ever considered including grapheme breaking for your library? 
This is needed for text editing to find out where the edges of graphemes. E.g. when two unicode characters are displayed as one grapheme. For example x plus Unicode Character 'COMBINING DIAERESIS' (U+0308) resulting in an x with two dots on top.

When the editor works only on unicode characters the user needs to press the cursor key twice to skip over the displayed text or hit the delete key twice to remove the text.

I do thing that is would fit thematically into this library.

Here is an example implementation
https://github.com/svarshavchik/courier-libs/blob/master/unicode/unicode_graphemebreak.c

**Comment by tasn** [comment]
In general it's a good idea and well within the scope of this library.

The reason why it was not added is because it was never asked for or needed by anyone, at least not for all we've been told.

**Comment by roever** [comment]
I could try to adapt the above library code for libunibreak... make the interface of courier-libs fit into libunibreak and so on and then provide a patch....
interested?

**Comment by tasn** [comment]
Please don't. If you'd like to add support to libunibreak, you'd have to write it from scratch (the algorithm is rather easy anyway).

The reason is that the above link is licensed under the GNU GPLv3 which is a copyleft license and would force us to relicense libunibreak to match.

**Comment by roever** [comment]
all right, didn't consider this, and yes the algorithm is simple... I have started. I would like some feedback though:
- the GB algorithm seems to not need language information, should I remove the language parameter?
- the current implementation will contain in brk[0] the information for in between the first two characters. The algorithm also includes information for in front of the first character... but I guess that that is not really interesting information. Do you think different? If we want to include this information brk must be one element longer than the text has characters... which is a bit counter intuitive

**Comment by adah1972** [comment]
If you do not see any chances of using the language information to customize the library behaviour, it will be unnecessary to have the language parameter.

I am just curious when the break information before the first character can be useful. Can you think of any?

**Comment by roever** [comment]
The language parameter question was asked because I thought you might want to have a similar interface to all algorithms.

And no I don't see a use for the breaking information in front of the first character. Especially as it will always be "break allowed" with the current algorithm... and I don't see that to change in future versions

**Comment by adah1972** [comment]
I would like to hear Tom's opinion on the lang parameter.

For the second part, definitely make it work like the current line break and word break functions. We do not want to have an interface that looks similar but is actually off by one, for no good reason!

**Comment by tasn** [comment]
I agree with @adah1972 regarding the second part, make it the same as the rest of the library.

As for the "lang" parameter: It seems that the standard implies it may be useful for some languages, and while it doesn't cover any language specific rules at the moment, it doesn't completely reject the notion. A quote from [the UAX](http://unicode.org/reports/tr29/#Grapheme_Cluster_Boundaries):

> The Unicode definitions of grapheme clusters are defaults: not meant to exclude the use of more sophisticated definitions of tailored grapheme clusters where appropriate. Such definitions may more precisely match the user expectations within individual languages for given processes. For example, “ch” may be considered a grapheme cluster in Slovak, for processes such as collation. The default definitions are, however, designed to provide a much more accurate match to overall user expectations for what the user perceives of as characters than is provided by individual Unicode code points. 

Though even without this paragraph, there are, in my opinion, two sufficient reasons why you'd want to have one:

1. Dealing with languages has taught me to always expect the unexpected. For every rule, there's a language somewhere that requires an exception.
2. Consistency with the other parts of the library.

In summary: have a lang parameter.

**Comment by roever** [comment]
agreed, will do so

**Comment by tasn** [comment]
I'm closing this one because there's already a PR for this. 

#18
